src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/PatternVariableAssignmentCheck.java

```diff
@@ -0,0 +1,118 @@
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+///////////////////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.checks.coding;
+
+import java.util.ArrayList;
+
+import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+
+public class PatternVariableAssignmentCheck extends AbstractCheck {
+
+  @Override
+  public int[] getRequiredTokens() {
+    return new int[]{TokenTypes.LITERAL_INSTANCEOF};
+  }
+
+  @Override
+  public int[] getDefaultTokens() {
+    return getRequiredTokens();
+  }
+
+  @Override
+  public int[] getAcceptableTokens() {
+    return getRequiredTokens();
+  }
+
+  @Override
+  public void visitToken(DetailAST ast) {
+
+    ArrayList<DetailAST> patternVariablesArray = getArrayOfPatternVariables(ast);
+
+    if (!patternVariablesArray.isEmpty()) {
+
+      DetailAST ifStatementSLIST = ast.getParent().getParent().findFirstToken(TokenTypes.SLIST);
+
+      for (DetailAST expressionToken = ifStatementSLIST.findFirstToken(TokenTypes.EXPR);
+           expressionToken != null; expressionToken = expressionToken.getNextSibling()) {
+
+        if (expressionToken.getType() == TokenTypes.EXPR) {
+          DetailAST assignToken = expressionToken.findFirstToken(TokenTypes.ASSIGN);
+
+          if (assignToken != null) {
+            String variableName = assignToken.findFirstToken(TokenTypes.IDENT).getText();
+
+            for (DetailAST patternVariableAST : patternVariablesArray) {
+              DetailAST patternVariableIdent = patternVariableAST.findFirstToken(TokenTypes.IDENT);
+
+              if (variableName.equals(patternVariableIdent.getText())) {
+                log(assignToken.getLineNo(), "Assignment of pattern variable is not allowed");
+                break;
+              }
+
+            }
+
+          }
+        }
+      }
+    }
+
+  }
+
+  /**
+    * Gets the list of all pattern variables in instanceof expression.
+    *
+    * @param ast ast tree of instanceoff to get the lsit from.
+    * @return list of pattern variables.
+    */
+  private ArrayList<DetailAST> getArrayOfPatternVariables(DetailAST ast) {
+    ArrayList<DetailAST> patternVariablesArray = new ArrayList<DetailAST>();
+
+    DetailAST outermostPatternVariable = ast.findFirstToken(TokenTypes.PATTERN_VARIABLE_DEF);
+    DetailAST recordPatternDef = (ast.getType() == TokenTypes.LITERAL_INSTANCEOF)
+        ? (ast.findFirstToken(TokenTypes.RECORD_PATTERN_DEF)) : ast;
+
+    if (outermostPatternVariable != null) {
+      patternVariablesArray.add(outermostPatternVariable);
+    }
+    else if (recordPatternDef != null) {
+      DetailAST recordPatternComponents = recordPatternDef
+          .findFirstToken(TokenTypes.RECORD_PATTERN_COMPONENTS);
+
+      for (DetailAST innerPatternVariable = recordPatternComponents.getFirstChild();
```

Comment:
```suggestion
      for (DetailAST innerPatternVariable = recordPatternDef
          .findFirstToken(TokenTypes.RECORD_PATTERN_COMPONENTS).getFirstChild();
```

Reply:
no long statements. extra variable is ok.